src/Microsoft.DotNet.VersionTools/Automation/GitHubApi/GitHubClient.cs

```diff
@@ -0,0 +1,211 @@
+                    email = _auth.Email
+                },
+                content = ToBase64(newFileContents),
+                sha = currentSha
+            }, Formatting.Indented);
+
+            var bodyContent = new StringContent(updateFileBody);
+            using (HttpResponseMessage response = await _httpClient.PutAsync(fileUrl, bodyContent))
+            {
+                response.EnsureSuccessStatusCode();
+                Trace.TraceInformation("Updated the file successfully.");
+            }
+        }
+
+        public async Task PostGitHubPullRequestAsync(
+            string title,
+            string description,
+            GitHubBranch headBranch,
+            GitHubBranch baseBranch)
+        {
+            string createPrBody = JsonConvert.SerializeObject(new
+            {
+                title = title,
+                body = description,
+                head = $"{headBranch.Project.Owner}:{headBranch.Name}",
+                @base = baseBranch.Name
+            }, Formatting.Indented);
+
+            string pullUrl = $"https://api.github.com/repos/{baseBranch.Project.Segments}/pulls";
+
+            var bodyContent = new StringContent(createPrBody);
+            using (HttpResponseMessage response = await _httpClient.PostAsync(pullUrl, bodyContent))
+            {
+                response.EnsureSuccessStatusCode();
+
+                Trace.TraceInformation($"Created pull request.");
+                Trace.TraceInformation($"Pull request page: {await GetPullRequestUrlAsync(response)}");
+            }
+        }
+
+        public async Task UpdateGitHubPullRequestAsync(
+            GitHubProject project,
+            int number,
+            string title = null,
+            string body = null,
+            string state = null)
+        {
+            var updatePrBody = new JObject();
+
+            if (title != null)
+            {
+               updatePrBody.Add(new JProperty("title", title));
+            }
+            if (body != null)
+            {
+                updatePrBody.Add(new JProperty("body", body));
+            }
+            if (state != null)
+            {
+                updatePrBody.Add(new JProperty("state", state));
+            }
+
+            string url = $"https://api.github.com/repos/{project.Segments}/pulls/{number}";
+
+            HttpRequestMessage request = new HttpRequestMessage(new HttpMethod("PATCH"), url);
+            request.Content = new StringContent(JsonConvert.SerializeObject(updatePrBody));
+
+            using (HttpResponseMessage response = await _httpClient.SendAsync(request))
+            {
+                response.EnsureSuccessStatusCode();
+
+                Trace.TraceInformation($"Updated pull request #{number}.");
+                Trace.TraceInformation($"Pull request page: {await GetPullRequestUrlAsync(response)}");
+            }
+        }
+
+        public async Task<GitHubPullRequest> FirstPullRequestWithHeadOrDefaultAsync(
```

Comment:
"OrDefault" was to match Linq naming: returns null rather than throwing if it's not found. Switching away from First gets rid of that odd expectation which is probably better. GitHub API docs call it a Search, so `SearchPullRequestsAsync` sounds good to me.


---
In reply to: [72436710](https://github.com/dotnet/buildtools/pull/919#discussion_r72436710) [](ancestors = 72436710)